app-backend/tool/src/main/scala/ast/MapAlgebraAST.scala

```diff
@@ -1,28 +1,26 @@
 package com.azavea.rf.tool.ast
 
+import geotrellis.raster.render.ColorRamp
+import io.circe.generic.JsonCodec
+
 import java.util.UUID
 
-import io.circe.generic.JsonCodec
 
 /** The ur-type for a recursive representation of MapAlgebra operations */
 sealed trait MapAlgebraAST extends Product with Serializable {
   def id: UUID
   def args: List[MapAlgebraAST]
   def label: Option[String]
-  def evaluable: Boolean
-  def unbound: List[MapAlgebraAST]
   def find(id: UUID): Option[MapAlgebraAST]
+  def sources: Seq[UUID]
+  def validateSources(sourceIds: Seq[UUID]): Boolean =
```

Comment:
Consider `Set` for fast lookups? Otherwise the `foldLeft` is `O(n^2)` in worst case.